### DISCUSSION TITLE
Tdarr with Hardware acceleration

### DISCUSSION THREAD

**Discussion by f-vicente** [tteck/Proxmox#1278] [Ideas]
Like Plex/Jellyfin/Emby, Tdarr uses ffmpeg backend to do transcoding and will take advantage of hardware acceleration.

**Comment by karl0ss** [comment]
I would love to see this, and having been playing with the Tdarr CT I have been unable to get it to work with my Intel GPU, but the installed plex CT (both from here) seems to use GPU for transcoding (confirmed by checking the dashboard and seeing (hw) ) so that defiantly works, so it feel like it shouldn't be too hard to do in the Tdarr CT? any idea/guidance @tteck

**Reply by tteck** [reply]
was the Tdarr LXC created privileged?

**Reply by f-vicente** [reply]
Is not too difficult to copy :

1. Install tdarr CT
2. copy the hardware acceleration config from your plex CT to the Tdarr CT .conf (on /etc/pve/lxc/XXX.conf) 
(on my install is as follows)
`lxc.cgroup2.devices.allow: c 226:0 rwm
lxc.cgroup2.devices.allow: c 226:128 rwm
lxc.mount.entry: /dev/dri/card0 dev/dri/card0 none bind,optional,create=file
lxc.mount.entry: /dev/dri/renderD128 dev/dri/renderD128 none bind,optional,create=file`
3. Start tdarr CT, add debian non-free apt repo, and install intel-media-va-driver-non-free drivers

**Reply by tteck** [reply]
1) Was it created privileged?  Default settings are unprivileged. Privileged is required!
2)
3) `apt-get install va-driver-all`  (non-free repository is not required)

It requires the installation of additional packages and commands from the `plex-install.sh`

**Reply by tteck** [reply]
I apologize for the confusion caused, as I mistakenly interchanged the conversations of @karl0ss and @f-vicente.

**Reply by f-vicente** [reply]
@tteck I have my privileged Tdarr CT working with non-free drivers, maybe was not needed, but others that I've tried didn't work.

I took commands and configuration from plex-install.sh as you suggest and transcoding works.

This discussion was only an idea to implement it by default, because it can be useful for others.

**Reply by tteck** [reply]
Now by default, the installation script of Tdarr provides hardware acceleration support.

**Reply by f-vicente** [reply]
Great! Thank you!

**Reply by karl0ss** [reply]
Amazing, I have just reinstalled CT via the script, and noticed that it is not enabling HWAccel, @f-vicente what profile have you used to have working passthough, I have an intel chip, trying to do qsv passthough, this seems to work on the plex CT, but still not working on this new CT, maybe I am missing something...

**Reply by f-vicente** [reply]
It’s a bit tricky because the complexity of tdarr first time configuration… I ended using a plugin called “Boosh-transcode using QSV GPU &FFMPEG”, but that might be the reason that caused it problems until I installed the private intel drivers. 

With the solution posted it should work with any of the vaapi transcoding plugins.

**Reply by karl0ss** [reply]
Interesting, I am also using the Bosh-transcode, I had to rebuild/install ffmpeg in that created CT and tell Tdarr to use that version of ffmpeg and QSV works, though, my whole promox host dies after about an hour and I have to reboot, works fine if using CPU, and seems to do a few files via GPU before the host dies, maybe I will try one of the vaapi plugins instead...(i'm quite new to tdarr)

**Comment by zchesler** [comment]
I can't get tdarr to work on the LXC with vaapi or other QSV plugin. no matter what I try and what I do, Please advise.
I'm running proxmox on N5105 intel celeron miniPC which I understand that has QSV capabilities.

just being clear - LXC Tdarr works. but no iGPU transcoding.

log attached.
[c2Lt2RVXp-log.txt](https://github.com/tteck/Proxmox/files/11495722/c2Lt2RVXp-log.txt)

Thanks

**Reply by tteck** [reply]
Is the container privileged?

**Reply by zchesler** [reply]
Yes.

**Reply by kloknibor** [reply]
I was fighting with this as well. If you try to to run certain CPU's you will need the non free library to support VAAPI. See: https://packages.debian.org/sid/intel-media-va-driver-non-free (this one is not installed by default)

**Reply by tteck** [reply]
The `va-driver-all` package installs the `intel-media-va-driver:amd64 21.1.1+dfsg1-1`, which is compatible with the Intel Core processor family (GEN8+) you mentioned. You can find more information about this package at https://packages.debian.org/bullseye/intel-media-va-driver. I strongly advise against adding the "unstable" repository unless it is absolutely necessary.

**Reply by f-vicente** [reply]
> The `va-driver-all` package installs the `intel-media-va-driver:amd64 21.1.1+dfsg1-1`, which is compatible with the Intel Core processor family (GEN8+) you mentioned. You can find more information about this package at https://packages.debian.org/bullseye/intel-media-va-driver. I strongly advise against adding the "unstable" repository unless it is absolutely necessary.

Most of people use Tdarr to encode old h264 to hvec. And the free driver states “HEVC and VP9 (Ice Lake and newer)”, that is gen10. Although using private drivers hvec is supported since gen7.

Could that be the problem?

**Reply by tteck** [reply]
🤷🏻‍♂️  sounds logical

**Comment by cbusillo** [comment]
It looks like there might be a few things missing for QSV on linux out of the box.  The built in ffmpeg on Tdarr is v4 and I think v5 is needed.  Debian 12 seems to work better as well.  It needs the non-free-firmware repo and these two packages installed: mkvtoolnix intel-media-va-driver-non-free.  The config needs to be changed to point to the system ffmpeg at /bin/ffmpeg.

Thank you for the great starting point, I think I have it working now.  If you want me to look into fixing these things, just let me know.  I've recently trudged through it, so I might remember most of it.  All of these things may not be needed, but it worked better for me.  Also sorry if some of this was addressed above, but I still had a hard time after reading through it :)

**Comment by Vetsus** [comment]
i know i'm probably fairly late to the game on this... but does the tteck script actually install a server AND node?  i'm only able to access a server.. maybe i'm missing something...

**Reply by Vetsus** [reply]
yeah... apparently i must have had something happen.. i deleted and reinstalled and it's working now

**Comment by littlesteve01** [comment]
Got anybody running HW Acceleration? I can´t get it working, tryed the whole night :-D 
Made yesterday a fresh install with the helper Scripts. 
Prozessor: Intel i7 6700T

with Jellyfin i got it working well but not for Tdarr. @tteck do you have any Idea? 

I installed the drivers via  `apt-get install va-driver-all `

**Reply by marc-cwm** [reply]
I've got one running with the exact same processor, Intel i7 6700T, and I'm also stumped at the moment.  I've got an i5 with the same generation in the same situation though.

**Reply by f-vicente** [reply]
For processors previous to gen 8 you need to install the proprietary drivers “intel-media-va-driver-non-free” if you want hvec encoding/decoding and use a qsv plugin in tdarr.

**Reply by tteck** [reply]
```
cat <<EOF >/etc/apt/sources.list

deb http://deb.debian.org/debian bookworm main contrib non-free non-free-firmware
deb-src http://deb.debian.org/debian bookworm main contrib non-free non-free-firmware

deb http://deb.debian.org/debian-security bookworm-security main contrib non-free non-free-firmware
deb-src http://deb.debian.org/debian-security bookworm-security main contrib non-free non-free-firmware

deb http://deb.debian.org/debian bookworm-updates main contrib non-free non-free-firmware
deb-src http://deb.debian.org/debian bookworm-updates main contrib non-free non-free-firmware
EOF

apt-get update
apt-get install -y intel-media-va-driver-non-free
```
